src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -0,0 +1,597 @@
+    }
+
+    /**
+     * Analyzes the current progress within the learning path and generates a recommended ordering of learning objects in a competency.
+     *
+     * @param learningPath the learning path that should be analyzed
+     * @param competency   the competency
+     * @param state        the current state of the recommendation
+     * @return the recommended ordering of learning objects
+     */
+    public List<LearningObject> getRecommendedOrderOfLearningObjects(LearningPath learningPath, Competency competency, RecommendationState state) {
+        var pendingLectureUnits = competency.getLectureUnits().stream().filter(lectureUnit -> !lectureUnit.isCompletedFor(learningPath.getUser())).toList();
+        ArrayList<LearningObject> recommendedOrder = new ArrayList<>(pendingLectureUnits);
+
+        // early return if competency can be trivially mastered
+        if (canBeMasteredWithoutExercises(competency)) {
+            return recommendedOrder;
+        }
+
+        final var combinedPriorConfidence = computeCombinedPriorConfidence(competency, state);
+        final var pendingExercises = competency.getExercises().stream().filter(exercise -> !learningObjectService.isCompletedByUser(exercise, learningPath.getUser()))
+                .collect(Collectors.toSet());
+        final var numberOfExercisesRequiredToMaster = predictNumberOfExercisesRequiredToMaster(learningPath, competency, combinedPriorConfidence, pendingExercises.size());
+        log.warn("numberOfExercisesRequiredToMaster:" + numberOfExercisesRequiredToMaster);
+        HashMap<DifficultyLevel, Set<Exercise>> difficultyLevelMap = generateDifficultyLevelMap(competency.getExercises());
+        if (numberOfExercisesRequiredToMaster >= competency.getExercises().size()) {
+            scheduleAllExercises(recommendedOrder, difficultyLevelMap);
+            return recommendedOrder;
+        }
+        final var recommendedExerciseDistribution = getRecommendedExerciseDistribution(numberOfExercisesRequiredToMaster, combinedPriorConfidence);
+        if (Arrays.stream(recommendedExerciseDistribution).sum() >= competency.getExercises().size()) {
+            scheduleAllExercises(recommendedOrder, difficultyLevelMap);
+            return recommendedOrder;
+        }
+
+        scheduleExercisesByDistribution(recommendedOrder, recommendedExerciseDistribution, learningPath, competency);
+        return recommendedOrder;
+    }
+
+    /**
+     * Checks if the competency can be mastered without completing any exercises.
+     *
+     * @param competency the competency to check
+     * @return true if the competency can be mastered without completing any exercises, false otherwise
+     */
+    private static boolean canBeMasteredWithoutExercises(@NotNull Competency competency) {
+        return ((double) competency.getLectureUnits().size()) / (3 * (competency.getLectureUnits().size() + competency.getExercises().size())) * 100 >= competency
```

Comment:
I've seen a variation of this formula and its derivatives in a few PRs now. Is it possible to move all related computations to one place, so it can be easily changed afterwards?